### ISSUE TITLE
Guard against implicit features

### ISSUE THREAD

**Issue by robertbastian** [unicode-org/icu4x#3004]
#2997 cleans up all implicit features (optional dependencies that generate features). We don't want those, as they're often unintended and unnecessarily increase the feature set.

**Comment by Manishearth** [comment]
We probably want to write a test here that uses `cargo metadata` or something to check all local crates and see if they have implicit features.

I do plan to be adding a crate under `tools` for https://github.com/unicode-org/icu4x/issues/2697, so this will probably eventually live there. For now, just make `tools/implicitfeatures` or something.

**Comment by psg-19** [comment]
Hi, i would like to work on this issue.
>For now, just make tools/implicitfeatures or something.

Is this still valid or can I just add a function that checks the `optional dependencies ` in `tools/make/depcheck/src/main.rs`, since it's related to dependency checking ?
I guess this was the crate @Manishearth was talking about and added it in this PR #3079

**Comment by robertbastian** [comment]
There has been some work in Cargo that would solve this since this issue was filed, but it seems to have been removed again: https://github.com/rust-lang/cargo/issues/12826, https://github.com/rust-lang/cargo/pull/14630. I think the next step here should be to get a clear picture of Cargo's status, i.e. is this available under nightly somehow? Can we get this back into Cargo?

**Comment by Manishearth** [comment]
Looks like they removed it from the edition because it was buggy and not going to be ready in time, and also they realized the right fix revolves around a deeper rearchitecting.

They don't seem to have made progress around the deeper rearchitecting. I can't see this being fixed soon, doesn't look like it's on nightly either.

**Comment by psg-19** [comment]
Yes, its not available in nightly.

>can I just add a function that checks the optional dependencies  in tools/make/depcheck/src/main.rs

so can I proceed with this ?

**Comment by psg-19** [comment]
Hi, I was working on it and faced few difficulties. 
1. In this [cargo.toml](https://github.com/unicode-org/icu4x/blob/main/components/experimental/Cargo.toml#L51) log is an implicit feature, but in `cargo metadata` output it somehow shows ` "dep:log"` and bypasses my implimentation of checking the implicit feature.


```rust
icu4x> cargo metadata --format-version 1 --no-deps | ConvertFrom-Json | Select-Object -ExpandProperty packages | Where-Object { $_.name -eq 'icu_experimental' } | Select-Object -ExpandProperty features


compiled_data : {dep:icu_experimental_data, icu_decimal/compiled_data, icu_list/compiled_data, icu_plurals/compiled_data...}
datagen       : {serde, dep:databake, zerovec/databake, zerotrie/databake...}
default       : {compiled_data}
log           : {dep:log}
ryu           : {fixed_decimal/ryu}
serde         : {dep:serde, zerovec/serde, potential_utf/serde, tinystr/serde...}



PS E:\shyam\icu4x> 
```

2. Should I keep something like hardcoded data of features which has to be implicit and should not be checked by the function ?

**Comment by Manishearth** [comment]
Looks like `cargo metadata` produces the  model of features with implicit features resolved, not the underlying data. Not quite sure what the best way to proceed is here, you probably could just parse the toml and get most of the info.

**Comment by psg-19** [comment]
Yes thats a great idea to parse the toml file, found a crate named [toml](https://crates.io/crates/toml) which can be used with `serde` to parse the toml data. Can I use it  for this purpose ?

**Comment by Manishearth** [comment]
Yes.
